rfcs/convergence/exposing-popper-options.md

```diff
@@ -0,0 +1,148 @@
+options that are required. This RFC proposes a way to configure popper uniformly in v9 components for consumers.
+
+## Background
+
+Configuring positioned props in both v0 and v8 are done inconsistently. Here are some examples
+
+v8:
+
+```tsx
+<Callout
+  {...props} // ICalloutProps interface
+/>
+
+<ContextualMenu
+  calloutProps={calloutProps} // ICalloutProps interface
+  // also can spread calloutProps
+  calloutMaxHeight
+  bounds
+  directionalHintFixed
+/>
+```
+
+v0:
+
+```tsx
+<Popup
+  align
+  position
+  offset
+/>
+
+<Menu
+  items={[
+    { menu: popper: { align, position, offset }}
+  ]}
+/>
+```
+
+## Problem statement
+
+There is a subset of `usePopper` options that are currently used by all converged components that need positioning.
+These options are spread into the component's props:
+
+```tsx
+<Tooltip position="bottom" align="center" />
+<Popover position="bottom" align="center" />
+<Menu position="bottom" align="center" />
+```
+
+There are other popper options that are used in some components but not others, examples of these options are:
+
+- `coverTarget` - allows the positioned element to cover the target
+
+There are also additional options that are not used by any converge component that might be used in the future like:
+
+- `autoSize` - Allows the positioned element to change size based on viewport size
+- `flipBoundary` - Determines what bounds to use to flip the positioning
+
+Issues and bugs can be triggered because certain use cases require more precise configuration of popper. The most likely
+result would be that a popper configuration is exposed for one component but not others.
+
+## Detailed Design or Proposal
+
+### Equal popper configuration
+
+Each converged component that uses popper should allow an equal level of configuration. This ensures that our positioned
+components can behave consistently across the library. Any fixes that require exposing another popper configuration will
+be available to **all** components that use popper.
+
+This proposal does not mean that all popper options must be public, but that if one component exposes a popper configuration,
+then all components should do so equally.
+
+We can achieve this with an agreement that the current `PositioningProps` would be supported by all v9 components, while
+internal popper options can still belong to `PopperOptions` type in the `react-positioning` package.
+
+```tsx
+import { PositioningProps } from '@fluentui/react-positioning';
+
+export interface TooltipProps extends PositioningProps {}
+
+export interface MenuProps extends PositioningProps {}
+
+export interface PopoverProps extends PositioningProps {}
+```
+
+### `positioning` configuration object
+
+The next step to this proposal is to make sure that all components which use a positioned element should allow users
+to configure popper through a `positioning` configuration object:
```

Comment:
I'm a little confused, in this RFC is the `positioning` object proposal intended to be an alternative to `extends PositioningProps`, or work together with it?

If it's either/or, then I have a slight preference to the `extends PositioningProps` option, since it makes the API for each control a little simpler:
`<Tooltip position="bottom" align="center" />`
instead of:
`<Tooltip positioning={{ position: 'bottom', align: 'center' }} />`

Reply:
it was intended to be an alternative that might be clearer for users (?) 

It's not that important and I can remove it for now, what's important is that we should be consistent across the library IMO

Reply:
There is a caveat with `extends PositioningProps`, it means that component authors will need to pass down all props to `usePopper()`:

```tsx
function Tooltip(props) {
  const {
    align,
    flipBoundary,
    overflowBoundary,
    popperRef,
    position,
    positionFixed,
    offset,
    unstable_disableTether,
    unstable_pinned,
    autoSize
  } = props;

  usePopper({
    align,
    flipBoundary,
    overflowBoundary,
    popperRef,
    position,
    positionFixed,
    offset,
    unstable_disableTether,
    unstable_pinned,
    autoSize
  });

  return <div />;
}
```

And usually, it's hard to track usages and it was so annoying that I created a test for it:

https://github.com/microsoft/fluentui/blob/31f949a20e204a266fc3849b128ea0be083c1dda/packages/fluentui/react-northstar/test/specs/commonTests/implementsPopperProps.tsx#L27-L41

⚠ The problem is that it works only with `Popper` component in v0, we don't have similar test for a hook.

---

Overall, it's one of reasons why I would prefer to use `positioning` as a single prop.

Reply:
That makes sense to me. What do you think about naming it `placement` instead of `positioning`?

Reply:
Also, I'm not 100% sure this is a great idea, but what about allowing a shorthand syntax for specifying position or align as a string:
```jsx
placement="above" // placement={{ position: 'above' }}
placement="top" // placement={{ align: 'top' }}
placement="below-start" // placement={{ position: 'below', align: 'start' }}
```

Reply:
- I think that we should merge `position` & `align` into a single property, `placement` makes sense
- Shorthand-like syntax may work 👍

Reply:
@behowell  I really like the shorthand idea 🚀 , will make it easy to distinguish between complex and simple configurations

Reply:
@behowell  @layershifter  I've added the proposal for the shorthand `placement` prop